### PR TITLE
Parsing parameter overhaul

### LINKED ISSUE #19119: ParsedMaterial arguments are not clear
## Reason
f_name is inherited from whichever module this material was moved from (phase field I m guessing)
It's hard to understand what s going on without looking at the docs, let s have better names

## Design
Rename some parameters, add old names as deprecated

## Impact
MOOSE user-friendliness

### PR DESCRIPTION
**Pull request by GiudGiud**
This is not necessarily the right names, but I d like to start the discussion with a working PR to show the extent of the effort required for this parameter rename. 

The motivation for this is that the framework uses different types for parsed inputs and the parameters are not clear as they currently stand. With a future uniformisation of parsed capabilities + adding parsed functors and functor arguments, this will only get worse. 

But the problem is that renaming them seems extremely disruptive to the phase field module

 thoughts about this
@dschwen @jiangwen84 

Closes #19119

### CODE REVIEW

**Comment by dschwen** [comment]
:-/

Lots of changes files. Does this change the args parameter everywhere? Is variable_names the best alternative? I suppose at least for parsed materials it mirrors material property names.

**Comment by GiudGiud** [comment]
This PR does change all the input files from args to variable_names.
I m not sure if variable_names is the best name. It s the best for ParsedMaterial for sure and that's where I determined it. 

In phase_field module, I think coupled_variables or parsed_variables is a slightly better name.

**Comment by hugary1995** [comment]
This discussion has actually happened more than once before. From what I remember, we couldn't find a much better alternative for `args`. I guess we didn't see what's particularly bad about `args`, hence we didn't have enough motivation to change this parameter.

**Comment by joshuahansel** [comment]
I've also been confused by parsed object param names. If it's open season on them, I'll suggest:
- `args` to `variable_names`
- `f_name` to `name` (unfortunately there is `material_property_names`; otherwise would call it `material_property_name`)
- `constant_expressions` to `constant_values` (why "expressions"? As I understand it, these would require `${fparse blah}` in expressions in the input file, so that they get parsed in the input file, not in the object, right?)
- `function` to `expression` (to distinguish from MOOSE Function)
- `value` to `expression`
- `vars`/`vals` is probably fine already, but maybe `expression_vars`/`expression_vals`, since `vars` can seem like coupled variables.

**Comment by dschwen** [comment]
> * `constant_expressions` to `constant_values` (why "expressions"? As I understand it, these would require `${fparse blah}` in expressions in the input file, so that they get parsed in the input file, not in the object, right?)

No, you can use _expressions_. This is also explained in the manual. Your constant expressions can use all constants declared prior in the same parameter vector!

**Comment by dschwen** [comment]
> * `f_name` to `name` (unfortunately there is `material_property_names`; otherwise would call it `material_property_name`)

The `f_name` stands for _function_ name. That's the name of the base material property only. In addition the material creates properties that are the derivatives, an those have generated name according to a fixed schema. I think simply calling this "material_property" is confusing.

**Comment by dschwen** [comment]
> * `vars`/`vals` is probably fine already, but maybe `expression_vars`/`expression_vals`, since `vars` can seem like coupled variables.

vars vals appears in `ParsedFunction`. I've always found that super confusing and just prepending `expression` does not clear anything up for me.

**Comment by joshuahansel** [comment]
> The f_name stands for function name. That's the name of the base material property only. In addition the material creates properties that are the derivatives, an those have generated name according to a fixed schema. I think simply calling this "material_property" is confusing.

"function name" seems confusing too. In light of the multiple properties, maybe `base_name`.

> vars vals appears in ParsedFunction. I've always found that super confusing and just prepending expression does not clear anything up for me.

Yeah, I was trying to communicate that `vars` are some tokens that appear in `expression` - they aren't MOOSE variables.

I don't feel strongly about any of my suggestions though, so I think I'd be happy as long as we're consistent, and the parameter descriptions no longer mention free energy 😛

**Comment by GiudGiud** [comment]
I took care of this last thing.
I do think if we rename parameters we should do it in one go, not make multiple PRs. 
`vars` is problematic I agree

`function_name` is an issue when the parsed object may depend on a function and have a `function_names` parameter

**Comment by GiudGiud** [comment]
how about constant_args instead of vars?

**Comment by GiudGiud** [comment]
miscellaneous_args ?

**Comment by GiudGiud** [comment]
This is still useful, though likely needs to be reduced in scope to be accepted

**Comment by GiudGiud** [comment]
how about `scalar_args / scalar_vals` instead of vars / vals?
for `constants, functions, PPs, and scalar variables`

@dschwen @joshuahansel

**Comment by GiudGiud** [comment]
so
args -> variable_names
f_name -> property_name (for ParsedMaterial) ?
constant_expressions still same
function -> expression
value -> expression
vars/vals ->scalar_vars/vals

**Comment by GiudGiud** [comment]
umm scalar_vars still isnt great for functions. Works for PPs, scalar variables and constants

**Comment by GiudGiud** [comment]
vars / vals
could be:
function_args (for vars)
argument_vals (for vals)

**Comment by joshuahansel** [comment]
> so
args -> variable_names
f_name -> property_name (for ParsedMaterial) base_name ?? ?
constant_expressions still same
function -> expression
value -> expression
vars/vals ->scalar_vars/vals

I like most of these. `args` could maybe just be `variables`, but `variable_names` works too. I'm still not sure about `f_name` - even though it's technically a sort of base name, I think my vote would be `property_name`. For `vars`/`vals`, I maybe like `args`/`vals`, (I think these parameters do not exist alongside the existing `args` parameter).

**Comment by dschwen** [comment]
`base_name` already has a different meaning and should not be used here. Anything with `scalar` screams `ScalarKernel` etc. and would be confusing, too. I do like `expression`, and for `args` I suggest `coupled_variables` (keep in mind that we use `args` in many places, among them Kernels and having both `variable` and `variables` as parameters would be ... suboptimal).

**Comment by GiudGiud** [comment]
I m worried about re-using args for vars/vals. I think re-using a deprecated param will be headaches. 
How about expression_args expression_args/vals? or arguments/values ?

```
args -> coupled_variables
f_name -> property_name (for ParsedMaterial)
constant_expressions still same
function -> expression
value -> expression
vars/vals -> args/vals

**Comment by joshuahansel** [comment]
I think we're close. Some other ideas:

- `expression`, `expression_args`/`expression_vals`
- `parsed_expression`, `parsed_args`/`parsed_vals`

I think `expression_vars` or `parsed_vars` would be fine too.

**Comment by GiudGiud** [comment]
I think it s poll time?

**Comment by dschwen** [comment]
Since the discussion apparently got split over github and slack, here's my slack comment.

Daniel Schwen
  [4 days ago](https://moosedevelopers.slack.com/archives/C01054VRUEM/p1666127570552439)
`vars` and `vals` are a ParsedFunction thing, and `f_name` is a `ParsedMaterial thing`. I suggested `coupled_variables` to replace `args`. For `ParsedFunction` all suggestions that retain `*vals` and `*vars` in the name are not an improvement, as they clarify nothing. (edited) 

Daniel Schwen
  [4 days ago](https://moosedevelopers.slack.com/archives/C01054VRUEM/p1666127769880799?thread_ts=1666127570.552439&cid=C01054VRUEM)
`symbol_names`, `symbol_values`

**Comment by GiudGiud** [comment]
Current list
```
args -> coupled_variables
f_name -> property_name (for ParsedMaterial)
constant_expressions still same
material_property_names still same
function -> expression
value -> expression
vars/vals -> symbol_names/symbol_values
```

**Comment by GiudGiud** [comment]
The new params added to the scope of changes made this an even bigger PR but it should be getting close to ready now

Other changes:
- The deprecation should be handled much more smoothly than in the initial version
- Fixed a couple heavy tests as I was running the test suite locally. It s odd that CIVET doesnt run them, I think the requisites are specific enough
- I tried to avoid changing all of phase field parameters this time. AC/CH etc remain with f_name

**Comment by GiudGiud** [comment]
#20535 was the wrong issue to use.
I ll change it to https://github.com/idaholab/moose/issues/19119 once this is full green

**Comment by GiudGiud** [comment]
Alright all green now.
I went for `coupled_variables` instead of `args`. Changing to `variable_names` is one `sed` away so if there are second thoughts, don't hold back.

**Comment by GiudGiud** [comment]
Added a getRenamedParam routine that comes in pretty handy for handling deprecation as a suggestion from @dschwen

**Comment by GiudGiud** [comment]
should be ready @dschwen

**Comment by GiudGiud** [comment]
@dschwen

**Review by dschwen — APPROVED** [review]

**Comment by GiudGiud** [comment]
Thanks!

**Comment by gambka** [comment]
I'm working on addressing the deprecation failures related to these changes in some downstream apps. Should the deprecation of `function` in `ParsedPostprocessor` in favor of `expression` also be made?

**Comment by GiudGiud** [comment]
There should be no failures, just warnings. 
It could be done. There wasnt as much confusion there.

**Comment by gambka** [comment]
> There should be no failures, just warnings. It could be done. There wasnt as much confusion there.

By failure I mean we have an allowed to fail test target that runs with `--error-deprecated`. When it is no longer green I prefer to make the necessary changes earlier rather than later to avoid users copying/using inputs still using the older syntax.

File: modules/phase_field/src/kernels/ADACInterface.C [hunk 1]
```diff
@@ -18,7 +18,12 @@ ADACInterface::validParams()
   params.addClassDescription("Gradient energy Allen-Cahn Kernel");
   params.addParam<MaterialPropertyName>("mob_name", "L", "The mobility used with the kernel");
   params.addParam<MaterialPropertyName>("kappa_name", "kappa_op", "The kappa used with the kernel");
-  params.addCoupledVar("args", "Vector of nonlinear variable arguments this object depends on");
+  params.addDeprecatedCoupledVar("args",
+                                 "Vector of nonlinear variable arguments this object depends on",
+                                 "args is deprecated, use 'coupled_variables' instead");
```
Comment (lindsayad) [inline modules/phase_field/src/kernels/ADACInterface.C:23]: Oh man this is what I get for introducing an API with a default argument. This is the API:
```
  /**                                                                                                                 
   * This method adds a deprecated coupled variable name pair.  The parser will look for variable                     
   * name pair in the input file and can return a reference to the storage location                                   
   * for the coupled variable if found. The doc string for the deprecated variable will be                            
   * constructed from the doc string for the new variable. A deprecation message will also be                         
   * automatically generated                                                                                          
   */                                                                                                                 
  void addDeprecatedCoupledVar(const std::string & old_name,                                                          
                               const std::string & new_name,                                                          
                               const std::string & removal_date = "");
```
This PR introduces ~40 wrong uses

Comment (lindsayad) [reply modules/phase_field/src/kernels/ADACInterface.C:23]: I will be deprecating this deprecation API

Comment (GiudGiud) [reply modules/phase_field/src/kernels/ADACInterface.C:23]: sounds good. 
Hope my PR isnt causing too many troubles

**Pull request merged**